### PR TITLE
🎉 New Source: Youtube analytics

### LINKED ISSUE #2970: New Source: Youtube Analytics
## Tell us about the new integration you’d like to have
https://developers.google.com/youtube/reporting/v1/reports 

## Describe the alternative you are considering or using

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368102600976) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by grubberr**
## What
#2970  - New Source: Youtube Analytics

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Connector added to connector index like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
   
</p>
</details>

### CODE REVIEW

**Review by bazarnov — CHANGES_REQUESTED** [review]
Tests are nicely done! Please consider to read the comments and make the corrections.
Keep up!

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        try:
+            self.get_jobs(authenticator)
+        except Exception as e:
+            return False, str(e)
+
+        return True, None
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        authenticator = self.get_authenticator(config)
+        jobs = self.get_jobs(authenticator)
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py]: https://developers.google.com/youtube/reporting/v1/reports#step-3:-create-a-reporting-job

File: airbyte-config/init/src/main/resources/seed/source_definitions.yaml [hunk 2]
```diff
@@ -595,3 +595,9 @@
   dockerImageTag: 0.1.3
   documentationUrl: https://docs.airbyte.io/integrations/sources/zuora
   sourceType: api
+- sourceDefinitionId: afa734e4-3571-11ec-991a-1e0031268139
+  name: YouTube Analytics
+  dockerRepository: airbyte/source-youtube-analytics
+  dockerImageTag: dev
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-config/init/src/main/resources/seed/source_definitions.yaml]: Done!

File: airbyte-integrations/connectors/source-youtube-analytics/acceptance-test-config.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  basic_read:
+    - config_path: "secrets/config.json"
+      configured_catalog_path: "integration_tests/configured_catalog.json"
+      empty_streams: []
+# TODO uncomment this block to specify that the tests should assert the connector outputs the records provided in the input file a file
+#      expect_records:
+#        path: "integration_tests/expected_records.txt"
+#        extra_fields: no
+#        exact_order: no
+#        extra_records: yes
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/acceptance-test-config.yml]: Done!

File: airbyte-integrations/connectors/source-youtube-analytics/acceptance-test-config.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    - config_path: "secrets/config.json"
+      status: "succeed"
+    - config_path: "integration_tests/invalid_config.json"
+      status: "failed"
+  discovery:
+    - config_path: "secrets/config.json"
+  basic_read:
+    - config_path: "secrets/config.json"
+      configured_catalog_path: "integration_tests/configured_catalog.json"
+      empty_streams: []
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/acceptance-test-config.yml]: Done!

File: airbyte-integrations/connectors/source-youtube-analytics/build.gradle [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    id 'airbyte-source-acceptance-test'
+}
+
+airbytePython {
+    moduleDirectory 'source_youtube_analytics'
+}
+
+dependencies {
+    implementation files(project(':airbyte-integrations:bases:source-acceptance-test').airbyteDocker.outputs)
+    implementation files(project(':airbyte-integrations:bases:base-python').airbyteDocker.outputs)
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/build.gradle]: Done!

File: airbyte-integrations/connectors/source-youtube-analytics/integration_tests/catalog.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
+{
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/integration_tests/catalog.json]: Done!

File: airbyte-integrations/connectors/source-youtube-analytics/integration_tests/configured_catalog.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
+{
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/integration_tests/configured_catalog.json:1]: Done!

File: airbyte-integrations/connectors/source-youtube-analytics/integration_tests/invalid_config.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/integration_tests/invalid_config.json:1]: Done!

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/schemas/channel_end_screens_a1.json [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+      "type": "string"
+    },
+    "end_screen_element_clicks": {
+      "type": "string"
+    },
+    "end_screen_element_impressions": {
+      "type": "string"
+    },
+    "end_screen_element_click_rate": {
+      "type": "string"
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/schemas/channel_end_screens_a1.json]: I have added "null"s for counter fields
but keep non-null for primary-key fields

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/spec.json [hunk 10]
```diff
@@ -0,0 +1,36 @@
+{
+  "documentationUrl": "https://docs.airbyte.io/integrations/sources/youtube-analytics",
+  "connectionSpecification": {
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "YouTube Analytics Spec",
+    "type": "object",
+    "required": ["client_id", "client_secret", "refresh_token"],
+    "additionalProperties": false,
+    "properties": {
+      "client_id": {
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/spec.json]: I have moved to "credentials" sub-object

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py]: every stream has to have one working "reporting job"
it's possible to create only one job for stream

all simple

if job - turned ON
stream produces daily data
if job is OFF 
stream empty

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream, HttpSubStream
+from airbyte_cdk.sources.streams.http.requests_native_auth import Oauth2Authenticator
+from airbyte_cdk.sources.utils.transform import TransformConfig, TypeTransformer
+
+
+class ReportResources(HttpStream):
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:60]: Added links

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        reports.sort(key=lambda x: x["startTime"])
+        date = kwargs["stream_state"].get("date")
+        if date:
+            reports = [r for r in reports if int(r["startTime"].date().strftime("%Y%m%d")) >= date]
+        return reports
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        if not self.job_id:
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:96]: job_id can be empty if nobody run "reporting job" before
if it's empty we try to run "reporting job" and get job_id

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+            client_secret=client_secret,
+            refresh_token=refresh_token,
+        )
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        authenticator = self.get_authenticator(config)
+        jobs_stream = JobsStream(authenticator=authenticator)
+
+        try:
+            jobs_stream.list()
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:163]: Are we receiving some sort of records here? If there are more than 1 records in the output, we can use `next()` to process only 1 record message, this should be enough for the check method.

**Review by bazarnov — COMMENTED** [review]
LGTM with small comments, please take a look.

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/google/YouTubeAnalyticsOAuthFlow.java [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.airbyte.config.persistence.ConfigRepository;
+import java.net.http.HttpClient;
+import java.util.function.Supplier;
+
+public class YouTubeAnalyticsOAuthFlow extends GoogleOAuthFlow {
+
+  public static final String SCOPE_URL = "https://www.googleapis.com/auth/yt-analytics.readonly";
```
Comment (alexandertsukanov) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/google/YouTubeAnalyticsOAuthFlow.java:16]: I believe this constant could be `private`. Do we use it outside the class?

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        if not current_stream_state:
+            return {self.cursor_field: latest_record[self.cursor_field]}
+        return {self.cursor_field: max(current_stream_state[self.cursor_field], latest_record[self.cursor_field])}
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return stream_slice["parent"]["downloadUrl"][len(self.url_base):]
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:133]:
```
pp stream_slice

{'parent': {'createTime': '2021-10-26T07:43:32.488579Z',
            'downloadUrl': 'https://youtubereporting.googleapis.com/v1/media/CHANNEL/ybpwL6sPt6SSzazIV400WQ/jobs/674aabf0-42e2-4b53-90c4-01f41f6f5ee4/reports/4315922755?alt=media',
            'endTime': '2021-09-26T07:00:00Z',
            'id': '4315922755',
            'jobId': '674aabf0-42e2-4b53-90c4-01f41f6f5ee4',
            'startTime': datetime.datetime(2021, 9, 25, 7, 0, tzinfo=datetime.timezone.utc)}}
```

I need to get "downloadUrl" and strip `url_base = 'https://youtubereporting.googleapis.com/v1/'` prefix
I am not sure how can I improve here 
?

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/google/YouTubeAnalyticsOAuthFlow.java (hunk 14 above)
Comment (grubberr) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/google/YouTubeAnalyticsOAuthFlow.java:16]: I am no sure, I am not java expert :)
I just copy-paste from this 
`./airbyte-oauth/src/main/java/io/airbyte/oauth/flows/google/GoogleAnalyticsOAuthFlow.java`

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        reports = []
+        for report in response_json.get("reports", []):
+            report = {**report}
+            report["createTime"] = datetime.datetime.strptime(report["createTime"], "%Y-%m-%dT%H:%M:%S.%f%z")
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py]: in really I don't remember why I converted "endTime", "createTime"
they are used nowhere 
I just removed them - to simplify code

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 13 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:161]: it make one HTTP request return one json response `{'jobs': [job1, job2, job3, jobN]}`
I cannot wrap `next` because it's not iterator 

https://github.com/airbytehq/airbyte/blob/c3330d257def9cd4fe5f199736213664b333240a/airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py#L43

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/schemas/channel_annotations_a1.json [hunk 17]
```diff
@@ -0,0 +1,51 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "properties": {
+    "date": {
+      "type": "integer"
+    },
+    "channel_id": {
+      "type": "string"
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/schemas/channel_annotations_a1.json]: I assume the slicing works from the `start_date` ?

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/google/YouTubeAnalyticsOAuthFlow.java (hunk 14 above)
Comment (grubberr) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/google/YouTubeAnalyticsOAuthFlow.java]: fixed

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:131]: Can we try this ? :
`url_base = ""`
```
def path(
        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
    ) -> str:
        return stream_slice["parent"]["downloadUrl"]
```

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/schemas/channel_annotations_a1.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/schemas/channel_annotations_a1.json]: for any type of iteration full refresh or incremental
this function `get_updated_state` will crash if 'date' start to come as 'null' values

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py]: yah, good point, improved

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return stream_slice["parent"]["downloadUrl"][len(self.url_base):]
+
+    def read_records(self, *, stream_slice: Mapping[str, Any] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+        parent = stream_slice.get("parent")
+        if parent:
+            yield from super().read_records(stream_slice=stream_slice, **kwargs)
+        else:
+            yield from []
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:138]: Can use provide some example ?
do you mean add some `self.logger.debug` ?

[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:138]: self.logger.info("....")

[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:139]: I have improved logging in that place

**Review by bazarnov — APPROVED** [review]

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+    def request_body_json(self, **kwargs) -> Optional[Mapping]:
+        if self.name:
+            return {"name": self.JOB_NAME, "reportTypeId": self.name}
+
+    def list(self):
+        " https://developers.google.com/youtube/reporting/v1/reference/rest/v1/jobs/list "
+        self.name = None
+        self.http_method = "GET"
+        results = list(self.read_records(sync_mode=None))
+        result = results[0]
```
Comment (midavadim) [inline airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:48]: check if result is not empty

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+
+    name = None
+    primary_key = "id"
+    url_base = "https://youtubereporting.googleapis.com/v1/"
+
+    def __init__(self, name: str, jobs_stream: JobsStream, job_id: str, **kwargs):
+        self.name = name
+        self.jobs_stream = jobs_stream
+        self.job_id = job_id
+        return super().__init__(**kwargs)
```
Comment (midavadim) [inline airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:71]: return is not needed

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+    name = None
+    primary_key = None
+    cursor_field = "date"
+    url_base = ""
+    transformer = TypeTransformer(TransformConfig.DefaultSchemaNormalization)
+
+    def __init__(self, name: str, dimensions: List[str], **kwargs):
+        self.name = name
+        self.primary_key = dimensions
+        return super().__init__(**kwargs)
```
Comment (midavadim) [inline airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:112]: return is not needed

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+        response_json = response.json()
+        reports = []
+        for report in response_json.get("reports", []):
+            report = {**report}
+            report["startTime"] = datetime.datetime.strptime(report["startTime"], "%Y-%m-%dT%H:%M:%S%z")
+            reports.append(report)
+        reports.sort(key=lambda x: x["startTime"])
+        date = kwargs["stream_state"].get("date")
+        if date:
+            reports = [r for r in reports if int(r["startTime"].date().strftime("%Y%m%d")) >= date]
```
Comment (midavadim) [inline airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:86]: this filter should be done on API side
https://developers.google.com/youtube/reporting/v1/reference/rest/v1/jobs.reports/list

Comment (midavadim) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:86]: Parameters:
    createdAfter

**Comment by midavadim** [comment]
Overall - please try to add more comments into code. It is not clear from the beginning why 'job' is needed, so it would be good to have more explanations

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 19 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:48]: yeah, we have already talked about this with @bazarnov 
it seems it's not very obvious for this stream class ( not sure how to make it more obvious )
but this stream always returns one and only one result because 
as you can see `parse_response` returns one and only one result.

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 20 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py]: yes, good catch, fixed

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 21 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py]: fixed

File: airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py (hunk 22 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-youtube-analytics/source_youtube_analytics/source.py:86]: Yes I know about this query param, but it's not what I need

This parameter filter on report `createTime` field but for main logic I need to use `startTime` field.
Report `startTime` and `createTime` are different things,
report can cover more earlier period than it was created.

Also YouTube don't  keep reports infinitely.
It will start to remove old reports,  number of reports will always be about 60 - 90.

**Review by midavadim — APPROVED** [review]

**Review by sherifnada — COMMENTED** [review]
Can you explain in a docstring the lifecycle of a job and how it relates to streams, as well as the hierarchy of different streams? It's hard to follow how everything works here. Specifically: 

1. Given that jobs can run for a very long time, is the connector just launching jobs and if there are past ready jobs, it reads the results from those? 
2. How does the connector "remember" the job IDs of previously launched jobs? or does it not need to because it uses the `list` endpoint? 
3. do we create one job per connector run, and that generates multiple reports? 

this is my pieceing together of what's happening based on reading the code. but it would be very valuable to write this all out explicitly to save the reader a lot of time

**Review by sherifnada — APPROVED** [review]
